plugins/user.py

```diff
@@ -0,0 +1,454 @@
+        seconds_since_last_verify = (now - self.verified_time).total_seconds()
+        return seconds_since_last_verify < self.session_duration
+
+    @property
+    def timezone(self):
+        """The user's timezone as a pytz.timezone()"""
+        if not hasattr(self, '_timezone'):
+            full_user_info = self.plugin.get_hipchat_user(self.hipchat_id)
+            self._timezone = pytz.timezone(full_user_info['timezone'])
+
+        return self._timezone
+
+    def localize_time(self, dt):
+        """Given a datetime object, localize it in the user's timezone"""
+        return dt.astimezone(self.timezone)
+
+    @property
+    def session_expiration_time(self):
+        """The localized datetime that the user's session expires or None"""
+        if self.verified_time is None:
+            return None
+
+        return self.localize_time(self.verified_time + datetime.timedelta(seconds=self.session_duration))
+
+    def logout(self):
+        """
+        Ensure that the user has to authenticate again once this action completes if they wish to continue the session.
+        """
+        self.verified_time = None
+
+    def save(self):
+        """
+        Persist the model in the plugin storage.
+
+        Note this encrypts the secure token before storing it externally.
+        """
+        as_dict = {
+            'token': encrypt(settings.TWOFACTOR_SECRET, self.token),
+            'permissions': self.permissions,
+        }
+        if self.verified_time:
+            as_dict['verified_time'] = self.verified_time
+        self.plugin.save(User.get_key(self.hipchat_id), as_dict)
+
+    def delete(self):
+        """Remove the user"""
+        self.plugin.clear(User.get_key(self.hipchat_id))
+
+    def generate_and_upload_qr_code_image(self):
+        """
+        Generate a QR code that can be used to configure Google Authenticator.
+
+        This URL can be accessed for 60 seconds by anyone who has it, so it should be handled with care.
+        """
+        s3_twofactor_bucket = settings.TWOFACTOR_S3_BUCKET
+        s3_twofactor_path_prefix = ''
+        parsed_url = urlparse.urlparse(s3_twofactor_bucket)
+        if parsed_url.scheme == 's3':
+            s3_twofactor_path_prefix = parsed_url.path
+            s3_twofactor_bucket = parsed_url.netloc
+
+        url = self.totp.provisioning_uri(self.nick, issuer_name=settings.TWOFACTOR_ISSUER)
+        img = qrcode.make(url)
+        conn = boto.connect_s3(profile_name=settings.TWOFACTOR_S3_PROFILE)
+        bucket = conn.get_bucket(s3_twofactor_bucket)
+        key = Key(bucket)
+        key.key = '{path}{name}-qr.png'.format(
+            path=s3_twofactor_path_prefix,
+            name=base64.urlsafe_b64encode(self.hipchat_id)
+        )
+        image_buffer = cStringIO.StringIO()
+        img.save(image_buffer, 'PNG')
+        key.set_contents_from_string(image_buffer.getvalue())
+        url = key.generate_url(10, query_auth=True)
```

Comment:
the '10' here provides a 10 second lifetime on the URL.  In my experience, tabbing out of a hipchat window unloads all images, and they're reloaded on tabbing back in.   In theory this will stop replay attacks.